soroban-env-host/src/native_contract/token/public_types.rs

```diff
@@ -60,6 +60,17 @@ pub enum Identifier {
     Account(AccountId),
 }
 
+impl Identifier {
+    pub(crate) fn to_val(self, env: &Host) -> Result<RawVal, Error> {
```

Comment:
This is probably not the right way to handle it. When I try to convert the Identifier into a RawVal/Object to be inserted into the topics, it gets converted to a Vector which disqualifies as a topic, thus I added this function. cc @leighmcculloch @sisuresh . Let me know if there's a better fix.

Reply:
Wow, this is a problem. We overlooked this, but we really need to be able to allow Vecs as topics because all custom types that are enums get translated into a Vec. So we at least need to allow small Vecs of 2 elements.

cc @graydon

Reply:
Also, custom types that are structs will be Maps, so we're also saying those types can't be topics. And note that many enums will be a 2 element Vec with a Map inside the second element for a struct.

Reply:
Would it be reasonable to remove the restriction on topic size/type, so that we can merge this with the Identifier being added as it normally would, and then we can create an issue for reintroducing the limitations in a way that allow for small Vec / Maps?

Reply:
I'm ok with that, but I think @sisuresh should be in a better position to comment.

Reply:
@graydon and Jon actually brought this up [here](https://github.com/stellar/stellar-protocol/issues/1298#issuecomment-1227613772). The topic restrictions keep you from putting a UDT into the topic. I'm not sure how helpful topics will be if they contain UDT's. The idea here was to allow downstream systems to make filtering easy with simple types. 

We could remove the restriction and implement a size restriction instead like mentioned in that issue to bypass this.

Reply:
@leighmcculloch How can I understand better how these UDTs are converted? I try to dig through the macro library but it confuses me.

Reply:
Yeah I'm fine with removing the limitations for now.

Reply:
Ok, I'll remove the limitations for now, seems that's the most straightforward fix, and then add an issue.

Reply:
I've got it on my todo to write a doc page for this, but in summary:

- Structs are encoded as `ScMaps`, where fields are `Symbol` keys, and the values are the type of the field.
- Enums that have unit and tuple variants are encoded as `ScVecs`, where the first element is a `Symbol` of the name of the variant, and the second element onwards are the tuple values if there are any.
- Enums with integer values are encoded as `U32`.

Reply:
Fixed and created https://github.com/stellar/rs-soroban-env/issues/534